rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,212 @@
+  extract: (state: PersistableState) => [PersistableState, SavedObjectReference[]]
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest state version with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const expressionReadyForSaving = persistableStatePlugin.beforeSave('expression', expressionString);
+
+``` 
+
+## EnhacedDrilldownEmbeddableInput
+
+Drilldown  plugin adds to the state of embeddable:
+
+```ts
+interface EnhancedDrilldownEmbeddableInput extends EmbeddableInput {
+  enhancements: {
+    drillDowns: { events?: Events[];  }
+  }
+}
+```
+Embeddable is aware that it has an enhancements property on its state, where every key represents state of another plugin.
+
+```ts
+const inject = (state, references) => {
+  Object.getKeys(state.enhancements).forEach(key => {
+    State.enhancements[key] = persistableStatePlugin.get(key).inject(state.enhancements[key], references);
+  }
+}
+```
+
+And drilldown plugin registers its state with persistableStateService.
+
+## We have to avoid shared persistable state
+
+So we have to do two things:
+
+- Always call a migration function for nested state.
+- Maintain state boundaries along plugin boundaries. Do not allow VisualizeEmbeddableInput to migrate data on EmbeddableInput. Don't even give it access to this state at all.
+As long as state ownership is isolated, we can do something like this:
+
+```ts
+const embeddableMigrator: PersistableStateMigrationFn (state: { type: string; input: unknown }, version: string): State {
+  return {
+    ...migrateInputToLatest(state.input),
+    enhancements: {
+       ...state.input.enhancements.map(enhancement => 
+          persistableStateMigrations.get(enhancement).migrate(state.input.enhancements[extension], version)),
+    },
+    specializedState: persistableStateMigrations.get(state.type).migrate(state.input.specializedState, version)),
```

Comment:
I'd be curious to see how the conversion from deserialized state to `EmbeddableInput` will go, and whether we can continue to assume users of Embeddables can still access things like:

```ts
const input: VisualizeEmbeddableInput = visEmbeddable.getInput();
console.log(input.vis.type);
```

vs

```ts
const input: VisualizeEmbeddableInput = visEmbeddable.getInput();
console.log(input.specializedState.visualize.type);
```

If we were able to have the serialized representation different from the deserialized representation, we could do this more formally.

```ts
const input: VisualizeEmbeddableInput = visEmbeddable.getInput();
console.log(input.vis.type);

const serializedState = visEmbeddablePlugin.beforeSave(input);
console.log(serializedState.specializedState.visualize.type);

// save to disk
// load from disk

const deserializedState = visEmbeddablePlugin.afterLoad(serializedState);
console.log(deserializedState.vis.type);
```

But the way the types are right now, deserialized and serialized state are the same.  Maybe that layer doesn't have to be part of this system, but if not, I'm not sure where it'd go. 🤔

Reply:
there is no serialization/deserialization happening inside this service

we are only working with Serializable state (which means a state that can be serialized, so no functions, circular references, .....)

this is kind of in between the StateInstance (EmbeddableInput instance) and SerializedState (for example json stored in database, or url string)

i dont know enough about embeddables, but for visualize state this goes something like this:

`vis` --> instance of Vis class, represents the state of each visualization
`vis.serialize()` --> returns VisSerializable
`persistableStateService.beforeSave('visualization', vis.serialize())` --> returns VisSerializable and references array (the actual type of the object should not change, we just extracted the references)
`savedVisLoader.save(id, persistableStateService.beforeSave('visualization', vis.serialize()))` --> saves to a saved object (i think the save method actually doesn't exist (yet, or in this form))

so with embeddable i could see something like EmbeddableInput, SerializableEmbeddableInput (we need some utility to convert between those) and the later is the one PersistableStatePlugin is interested in (and can be later serialized and stored)

Reply:
@Dosant can you take a look into this ?